File: Javascript/518. Coin Change 2.js
/**
 * Link: https://leetcode.com/problems/coin-change-2/
 * Authoir: Dung Nguyen Tien (Chris)
 * @param {number} amount
 * @param {number[]} coins
 * @return {number}
 */
var change = function(amount, coins) {
    var dp = Array(amount + 1).fill(0);
    dp[0] = 1;
    
    // put each coin in the amount
    // we only consider one coin once, so it wont repeat
    for (var coin of coins) {
        for (var i = coin; i <= amount; i++) {
            dp[i] += dp[i-coin];
        }
    }

    return dp[amount];
};

File: C#/518. Coin Change 2.cs
/*
 * Link: https://leetcode.com/problems/coin-change-2/submissions/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public int Change(int amount, int[] coins) {
        var dp = new int[coins.Length + 1][];
        for (var i = 0; i <= coins.Length; i++) dp[i] = new int[amount + 1];
        
        dp[0][0] = 1;
        for (var i = 1; i <= coins.Length; i++) {
            dp[i][0] = 1;
            
            for (var j = 1; j <= amount; j++) {
                dp[i][j] = dp[i - 1][j] // not pick coin i
                    + (j - coins[i - 1] >= 0 ? dp[i][j - coins[i - 1]] : 0); // pick coins i
            }
        }
        
        return dp[coins.Length][amount];
    }
}